src/detect-engine-build.c

```diff
@@ -1125,9 +1126,341 @@ static int RuleSetWhitelist(Signature *s)
     return wl;
 }
 
-int CreateGroupedPortList(DetectEngineCtx *de_ctx, DetectPort *port_list, DetectPort **newhead,
-        uint32_t unique_groups, int (*CompareFunc)(DetectPort *, DetectPort *));
-int CreateGroupedPortListCmpCnt(DetectPort *a, DetectPort *b);
+static int CreateGroupedPortListCmpCnt(const DetectPort *a, const DetectPort *b)
+{
+    if (a->sh->init->score < b->sh->init->score) {
+        return 1;
+    } else if (a->sh->init->score > b->sh->init->score) {
+        return -1;
+    }
+
+    if (a->sh->init->sig_cnt < b->sh->init->sig_cnt) {
+        return 1;
+    } else if (a->sh->init->sig_cnt > b->sh->init->sig_cnt) {
+        return -1;
+    }
+
+    /* Hack to make the qsort output deterministic across platforms.
+     * This had to be done because the order of equal elements sorted
+     * by qsort is undeterministic and showed different output on BSD,
+     * MacOS and Windows. Sorting based on id makes it deterministic. */
+    if (a->sh->id < b->sh->id) {
+        return -1;
+    } else {
+        return 1;
+    }
+}
+
+static int SortCompare(const void *a, const void *b)
+{
+    const DetectPort *pa = *(const DetectPort **)a;
+    const DetectPort *pb = *(const DetectPort **)b;
+
+    return CreateGroupedPortListCmpCnt(pa, pb);
+}
+
+static inline void SortGroupList(uint32_t *groups, DetectPort **list,
+        int (*CompareFunc)(const DetectPort *, const DetectPort *))
+{
+    int cnt = 0;
+    for (DetectPort *x = *list; x != NULL; x = x->next) {
+        BUG_ON(x->port > x->port2);
+        cnt++;
+    }
+    if (cnt <= 1)
+        return;
+
+    /* build temporary array to sort with qsort */
+    DetectPort **array = (DetectPort **)SCCalloc(cnt, sizeof(DetectPort *));
+    if (array == NULL)
+        return;
+
+    int idx = 0;
+    for (DetectPort *x = *list; x != NULL;) {
+        /* assign a temporary id to resolve otherwise equal groups */
+        x->sh->id = idx + 1;
+        SigGroupHeadSetSigCnt(x->sh, 0);
+        DetectPort *next = x->next;
+        x->next = x->prev = x->last = NULL;
+        BUG_ON(x->port > x->port2);
+        array[idx++] = x;
+        x = next;
+    }
+    BUG_ON(cnt != idx);
+
+    qsort(array, idx, sizeof(DetectPort *), SortCompare);
+
+    /* rebuild the list based on the qsort-ed array */
+    DetectPort *new_list = NULL, *tail = NULL;
+    for (int i = 0; i < idx; i++) {
+        *groups += 1;
```

Comment:
whats the point of this? This function doesn't modify the list length

Reply:
I think it existed for cases when there aren't unique groups received from the client or server settings, in which case, we assume, all the groups are a part of joingroup.
But, now, we don't need to update it in the loop each time, we can set it to the list len at the end.